### ISSUE TITLE
fullTemplateCheck not recognizing unary (+) operator in template

### ISSUE THREAD

**Issue by serbboy23** [angular/angular#36178]
# 🐞 bug report

### Description

I'm finally getting around to using ```angularCompilerOptions``` and have started to use ```fullTemplateCheck```. It seems as if while checking the templates, there is a failure to see the ```+``` unary operator in front of the ```hours``` variable (```hours``` is a string). The ```+``` turns the string into a number so I believe this is an incorrect error, and therefore a bug?

## 🔥 Exception or Error
<pre>
error TS2362: The left-hand side of an arithmetic operation must be of type 'any', 'number', 'bigint' or an enum type.
&ltdiv (click)="formatHours(<b>+hours</b> + 1)"&gt
</pre>

## Environment

**Angular Version:**
<pre><code>
Angular CLI: 9.0.4
Node: 10.16.1
OS: linux x64

Angular: 9.0.4
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.4
@angular-devkit/build-angular     0.900.4
@angular-devkit/build-optimizer   0.900.4
@angular-devkit/build-webpack     0.900.4
@angular-devkit/core              9.0.4
@angular-devkit/schematics        9.0.4
@angular/cdk                      9.1.0
@angular/material                 9.1.0
@ngtools/webpack                  9.0.4
@schematics/angular               9.0.4
@schematics/update                0.900.4
rxjs                              6.5.4
typescript                        3.7.5
webpack                           4.41.2
</code></pre>

**Comment by JoostK** [comment]
Thanks for the report! This is due to completely unexpected parser behavior, it would seem.

Apparently we there's no unary operators represented in the internal parse tree at all. Instead, the parser converts it to a binary operator:

- `+total` is interpreted as `total - 0`
- `-total` is interpreted as `0 - total`

WOW! 😕

This is similar to #20845 but I'd say its effect is a bit worse.

**Comment by JoostK** [comment]
The error for `displayHours == 23` is purely TypeScript related, and it appears to be intentional: microsoft/TypeScript#30655

**Comment by serbboy23** [comment]
Thank you for the quick response :smile: 
For the unary issue: I guess I could just use a method on the class. Is this something that might be looked into in the near future? I see the other issue is from 2017.

For the typescript issue: Wow ... interesting

**Comment by serbboy23** [comment]
My initial thought with the ```==``` was to use a ```+``` in front of the ```displayHours``` variable so that two `number`s are being compared, however that brings me back to the main issue at hand here.

**Comment by JoostK** [comment]
I can sympathize with TypeScript's choice here. Very unexpected things can happen with these sorts of comparisons. For instance, `1000 == '1.e3'` is `true` in JavaScript ;) You'd better off by converting the number to a string instead, avoiding the problem.

I edited my original comment for the unary operator with an alternative workaround to a method on the class. Your choice on which you choose.

I don't think this issue will be fixed soon to be honest, but let's hope it won't be 3 years ;) The occurrence of the issue from 2017 is less common.

**Comment by serbboy23** [comment]
Yeah the more I think about it, the more I realize there can be many unexpected results from allowing that loose comparison.
Here's to 2 years and 11 months to fix the issue haha! :beer:

**Comment by djleonskennedy** [comment]
@serbboy23 you're add all logic in template in your code ?
